### PR TITLE
feat: add node: prefixed modules

### PR DESCRIPTION
**Pull request by dnalborczyk**
**What kind of change does this PR introduce?**

follow-up from: https://github.com/webpack/webpack/pull/12181

adds 'node:' prefixed builtin modules.

https://nodejs.org/docs/latest-v12.x/api/esm.html#esm_node_imports

https://github.com/nodejs/node/pull/37246

**Did you add tests for your changes?**

**Does this PR introduce a breaking change?**

**What needs to be documented once your changes are merged?**

### CODE REVIEW

**Comment by alexander-akait** [comment]
Can you add small test?

File: lib/node/NodeTargetPlugin.js [hunk 1]
```diff
@@ -57,6 +57,9 @@ const builtins = [
 	"zlib"
 ];
 
+// add node: prefixed modules
+builtins.push(...builtins.map(builtin => `node:${builtin}`));
```
Comment (sokra) [inline lib/node/NodeTargetPlugin.js:61]:
```suggestion
	"zlib",
	/^node:/
];
```

**Review by sokra — CHANGES_REQUESTED** [review]

**Comment by sokra** [comment]
make to sure to also add a test case for that.

**Comment by sokra** [comment]
hmm, seems like `node:` only works in `import` and not in `require`, but we always generate `require` calls for the node.js externals. We need to revisit that when ESM support allows module externals.

**Comment by dnalborczyk** [comment]
> hmm, seems like `node:` only works in `import` and not in `require`, but we always generate `require` calls for the node.js externals. We need to revisit that when ESM support allows module externals.

@sokra

that's what I tried to explain here: https://github.com/webpack/webpack/pull/12181#issuecomment-745305067

this will be supported soon in node.js for commonjs modules as well (mentioned on top): https://github.com/nodejs/node/pull/37246

I thought **esm** support is currently being added to webpack? https://github.com/webpack/webpack/releases/tag/v5.22.0

**Comment by alexander-akait** [comment]
@dnalborczyk I think we need wait when it was merged and we can merge it too

**Comment by dnalborczyk** [comment]
> @dnalborczyk I think we need wait when it was merged and we can merge it too

@alexander-akait it's merged now. it seems it's being shipped in node v16.x and possibly backported later on.

**Comment by sokra** [comment]
Ok node merged `node:` support for require, so this PR is fine now. A test case is still missing for that.

**Comment by karlhorky** [comment]
This will close #13290 when merged

**Comment by alexander-akait** [comment]
@dnalborczyk friendly ping

**Comment by karlhorky** [comment]
So this feature is about to get more usage: the `node:` scheme / prefix / protocol was backported to Node.js 14.13.1 / 12.20.0 🎉

https://nodejs.org/api/esm.html -> link "node: Imports" (links with anchors don't scroll properly right now)

Screen Shot 2021-05-30 at 11 04 39

**Comment by karlhorky** [comment]
An alternative to the approaches in #13311 and #12693 would be to use [`is-core-module`](https://github.com/inspect-js/is-core-module) (which was split from [`resolve`](https://github.com/browserify/resolve) in [browserify/resolve@7c26483](https://github.com/browserify/resolve/commit/7c26483)), so that this list doesn't need to be maintained and tested by `webpack` itself.

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Comment by karlhorky** [comment]
Thanks @sokra! Looking forward to using this! (also with our students in Next.js!)

Will this be in a new patch (`5.39.2`) or minor (`5.40.0`)?

**Comment by fregante** [comment]
I'm still seeing this issue on webpack 5.52

```
ERROR in node:path
Module build failed: UnhandledSchemeError: Reading from "node:path" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.

[log clipped]

webpack 5.52.0 compiled with 1 error in 34836 ms
```

**Comment by alexander-akait** [comment]
@fregante please create steps to reproduce, if you use `target: 'web'` it is expected, no `node:path` in browser

**Comment by fregante** [comment]
Gotcha, I am using a web target. Is there no solution to this in webpack currently? https://npms.io/search?q=node+scheme+webpack

**Comment by alexander-akait** [comment]
You can use alias and use https://github.com/browserify/path-browserify

**Comment by alexander-akait** [comment]
But maybe better to ask developer provide browser based version of library

**Comment by fregante** [comment]
The developer does, but it uses [`exports.browser` in the package.json](https://github.com/sindresorhus/filenamify/blob/43f3f32743df4c9230749a29b00f700ef4e93878/package.json#L16), which doesn't seem to be picked up and I wasn't able to find any issues regarding exports map on this repo either.

I'm using an alias in webpack for this but it's not a tenable solution if more packages come along.

**Comment by fregante** [comment]
> You can use alias and use [browserify/path-browserify](https://github.com/browserify/path-browserify)

That's already configured. Are you saying that it can be handled completely by adding a specific `node:path` alias?

**Comment by alexander-akait** [comment]
> The developer does, but it uses exports.browser in the package.json, which doesn't seem to be picked up and I wasn't able to find any issues regarding exports map on this repo either.

It will be used if it is provided

**Comment by fregante** [comment]
> > You can use alias and use [browserify/path-browserify](https://github.com/browserify/path-browserify)
> 
> That's already aliased to `path` via [Richienb/node-polyfill-webpack-plugin](https://github.com/Richienb/node-polyfill-webpack-plugin). Are you saying that it can be handled completely by adding a specific `node:path` alias?

Answering myself, it doesn't work, same problem as before:

```js
module.exports = {
  target: "web",
  resolve: {
    fallback: {
      fs: false,
      crypto: false,
      console: false,
      vm: false,
      path: false,
      "node:path": false,
    },
    alias: {
      'node:path': "path-browserify",
    },
  },
```

**Comment by alexander-akait** [comment]
Give me time, I will investigate

**Comment by fregante** [comment]
I don't understand why you closed the issue. That's not a duplicate. I gave a full report including a demo repository avoiding to clutter an unrelated PR and you just locked it 🤷‍♂️

**Comment by fregante** [comment]
They're two separate issues:

- node:path, being discussed here
- exports.browser, for which I opened https://github.com/webpack/webpack/issues/14163

**Comment by alexander-akait** [comment]
Because the problem in other place, what is the problem wait the answer?

**Comment by alexander-akait** [comment]
Change your import on `import * as mod from "filenamify/browser";`

**Comment by alexander-akait** [comment]
Here described https://nodejs.org/api/packages.html#packages_subpath_exports

**Comment by alexander-akait** [comment]
And we cannot analyze it is subpath for browser or not, so you should explicitly use `filenamify/browser`

**Comment by alexander-akait** [comment]
But here other problem `resolve.fallback.path`/`resolve.fallback['node:path']` is not respected

**Comment by alexander-akait** [comment]
Issue for `resolve.fallback` and `resolve.alias` https://github.com/webpack/webpack/issues/14166

**Comment by fregante** [comment]
https://github.com/webpack/webpack/issues/14163 is showing up in search results. 

Please unlock the issue and explain that "browser" in the export map is not expected to work like the ["browser" field](https://github.com/defunctzombie/package-browser-field-spec). The current locking reason makes no sense.

**Comment by alexander-akait** [comment]
I don't know what we need explain here, Node.js docs describe how `exports` works

**Comment by fregante** [comment]
I’m stumped by your reasoning here. Webpack users will bump into that exact issue and there’s no explanation on it. It just does not work. Please be kind and add an explanation there, it’s free.

**Comment by fregante** [comment]
https://xkcd.com/979/

**Comment by alexander-akait** [comment]
Where we should docs it?

**Comment by fregante** [comment]
In the issue itself

**Comment by alexander-akait** [comment]
Here answer https://github.com/webpack/webpack/pull/12693#issuecomment-914258583, `exports` is just subpath of URL, we can't understand that code for browser or for node or for other env

**Comment by fregante** [comment]
I think you're completely missing the point.

1. This PR is unrelated
2. https://github.com/webpack/webpack/issues/14163 is the issue that's missing a solution, and is not linked to this one
3. I'm not the one needing a solution anymore

**Comment by alexander-akait** [comment]
> exports.browser in package.json is not automatically picked up on web target #14163 is the issue that's missing a solution, and is not linked to this one

And should not do it - `./browser` is part of URL, if you need special target change it on `browser`, i.e.:
```
{
	"exports": {
		"browser": "./browser.js",
		"default": "./index.js"
	}
}
```

No `./`

**Comment by fregante** [comment]
…

Can you just unlock that issue? There's nothing to discuss here.

**Comment by alexander-akait** [comment]
No need discussion, what is the problem? I answered the same multiple times, `./browser` will not and should not work as you expected, logic described in Node.js docs

**Pull request merged**